src/kernel/tcap.c

```diff
@@ -215,6 +218,30 @@ tcap_merge(struct tcap *dst, struct tcap *rm)
 	return 0;
 }
 
+int
+tcap_wakeup(struct tcap *tc, tcap_prio_t prio, struct thread *thd, struct cos_cpu_local_info *cli)
+{
```

Comment:
Yes, It makes sense. `cos_thd_wakeup` as it is now, doesn't control the next timer interrupt, which is one main difference from `cos_switch` and can be a limitation as you suggested. 
Implementing it could be simple, bookkeeping the budget in `next_thdinfo` and converting it to timeout(`tcap_time_t`) when `awoken thread` it is activated from `cos_rcv`.